calyx/src/passes/compile_sync.rs

```diff
@@ -0,0 +1,380 @@
+            //          generate the individual groups
+            //          build the seq stmt
+            let mut stmts_new: Vec<ir::Control> = Vec::new();
+            for stmt in seq.stmts.drain(..) {
+                if let Some(n) = stmt.get_attributes().unwrap().get("sync") {
+                    if map.get(n).is_none() {
+                        add_shared_structure(builder, n, map);
+                    }
+                    let (cells, groups) = map.get(n).unwrap();
+                    let barrier = Rc::clone(&cells[0]);
+                    let eq = Rc::clone(&cells[1]);
+                    let wait_restore = Rc::clone(&groups[0]);
+                    let restore = Rc::clone(&groups[1]);
+                    let clear_barrier = Rc::clone(&groups[2]);
+                    let n_new = *(count.entry(*n).or_insert(0)) + 1;
+                    count.insert(*n, n_new);
+                    structure!(builder;
+                            let save = prim std_reg(32););
+                    let incr_barrier = builder.add_group("incr_barrier");
+                    build_incr_barrier(
+                        builder,
+                        &incr_barrier,
+                        &barrier,
+                        &save,
+                        &(&n_new - 1),
+                    );
+                    let write_barrier = builder.add_group("write_barrier");
+                    build_write_barrier(
+                        builder,
+                        &write_barrier,
+                        &barrier,
+                        &save,
+                        &(&n_new - 1),
+                    );
+                    let wait = builder.add_group("wait");
+                    build_wait(builder, &wait, &eq);
+                    if n_new == 1 {
+                        stmts_new.push(build_member(
+                            &stmt,
+                            incr_barrier,
+                            write_barrier,
+                            wait,
+                            wait_restore,
+                        ));
+                    } else {
+                        stmts_new.push(build_member_0(
+                            &stmt,
+                            incr_barrier,
+                            write_barrier,
+                            wait,
+                            clear_barrier,
+                            restore,
+                        ));
+                    }
+                } else {
+                    stmts_new.push(stmt);
+                }
+            }
+            seq.stmts = stmts_new;
+        }
+        ir::Control::While(w) => {
+            build_barriers(builder, &mut w.body, map, count);
+        }
+        ir::Control::If(i) => {
+            build_barriers(builder, &mut i.tbranch, map, count);
+            build_barriers(builder, &mut i.fbranch, map, count);
+        }
+        _ => {}
+    }
+}
+
+fn build_incr_barrier(
```

Comment:
Make this method construct the `incr_barrier` group and return it instead of taking the `RRC<ir::Group>` reference to it.